From 0be3014cfc43cebbb9e826d0feb106ec6a6a67c4 Mon Sep 17 00:00:00 2001
From: William Guilen <wguilen@outlook.com>
Date: Wed, 28 Nov 2018 11:19:32 -0200
Subject: [PATCH] Fix conversion to KML when LineString should be a LinearRing

---
 lib/adapters/KML.class.php | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/lib/adapters/KML.class.php b/lib/adapters/KML.class.php
index 8be9f68..0730a21 100644
--- a/lib/adapters/KML.class.php
+++ b/lib/adapters/KML.class.php
@@ -224,6 +224,16 @@ private function pointToKML($geom) {
   private function linestringToKML($geom, $type = FALSE) {
     if (!$type) {
       $type = $geom->getGeomType();
+
+      if ($type === 'LineString') {
+        $components = $geom->getComponents();
+        $firstPoint = array_shift($components);
+        $lastPoint = array_pop($components);
+
+        if ($firstPoint->equals($lastPoint)) {
+          $type = 'LinearRing';
+        }
+      }
     }
 
     $str = '<'.$this->nss . $type .'>';
